### DISCUSSION TITLE
Translations

### DISCUSSION THREAD

**Discussion by i-make-robots** [MarginallyClever/Makelangelo-software#392] [Ideas]
IMHO if a translated string is only used in one class it should be `ClassName.KeyName`

This way a script can be written which scans all translation files and checks that the class exists and if the string is used in that class.  This prevents build up of old strings.

In the class code it can be written as ClassName.class.getSimpleName()+".KeyName".  This way if the class name changes the strings break (previous test mentioned).

It should also be possible then to write a test that all keys in english have matching KeyNames in the other languages.

**Comment by PPAC37** [comment]
Can someone explain / detail / give a context of use?

src/main/resources/languages/piglatin.xml 

Is it a dialect or a basis for translation into a language based on ideograms? (or maybe some kind of language like "Klingon" (Star Trek language)?)

Because I can't find it in https://www.loc.gov/standards/iso639-2/php/code_list.php

**Comment by PPAC37** [comment]
To support the discussion

We may need a tool to

* maintain a list of renamed '```<key>```' to be able to easily integrate/update a new translation file created on an old version and impact all the existing translation files. (or to maintain only english.xml and be able to automatically update the renamed  ```<key>``` (when we have the tool) in the other existing translation files. )
* add a timestamp or a version in the existing .xml translation files.
* add a pattern like "codeLang_codeCountry" ig "en_EN" or "fr_FR" (edit : at the end of the filename) to avoid loading all the .xml in the current directory when we load the language in the current dir of the user.

Modify the current lnguage .xml format so that the key is used as id and therefore uniqueness is easily checked / validated.
(Edit : but this is not actualy posible with a xsl based trasformation like src/main/resources/languages/language_v0_To_v1.xsl  in  https://github.com/MarginallyClever/Makelangelo-software/pull/455 due to the fact that some key value are not NCName compatible ( NCName as a name in xml ... have limitations ...  https://stackoverflow.com/questions/1631396/what-is-an-xsncname-type-and-when-should-it-be-used )

https://github.com/MarginallyClever/Makelangelo-software/pull/455
https://github.com/MarginallyClever/Makelangelo-software/pull/456

**Comment by PPAC37** [comment]
To ilustrat XML validation (under NetBeans 12.6 but you may have the equivalent under Eclips or IntelliJ )
(NO DTD -> errors )
![image](https://user-images.githubusercontent.com/94939582/149738219-4b6fe957-60da-4f9e-a19d-44bab2818973.png)
( With an adapted DTD )
![image](https://user-images.githubusercontent.com/94939582/149738283-9f990018-af41-4d64-bf66-f5f575777004.png)
on german.xml :
![image](https://user-images.githubusercontent.com/94939582/149739925-d207c60b-0582-4437-b2e7-0efdbde20168.png)

XML Transformation with a .xsl to add as id attribut to the element string the value of is element key
![image](https://user-images.githubusercontent.com/94939582/149738526-9e0a6c7e-af0a-4258-a23e-55bb5bb8d31e.png)

XML Validation of the output ( we no have a id unicity check cf "... must be unique ... " (but have to clean the key to be NCName (charlimitation like no space, ...) cf "... must be an NCName ... " )
![image](https://user-images.githubusercontent.com/94939582/149738978-8a6f8d5e-b807-4650-8d6a-5af821974582.png)

**Comment by PPAC37** [comment]
https://github.com/MarginallyClever/Makelangelo-software/issues/461
> in case of Duplicated Key with not the same value ?

https://github.com/MarginallyClever/Makelangelo-software/issues/461#issuecomment-1017766000
> Make a judgement call after asking a native speaker of the language which translation is more accurate.
> Also check if more than one place in the code uses the same phrase, it might be better to split into two keys.

---
https://github.com/MarginallyClever/Makelangelo-software/pull/462
> do we keep unused key ?

https://github.com/MarginallyClever/Makelangelo-software/pull/462#issuecomment-1017758632
> I wouldn't keep old keys. if we need to restore them ... that's what github is for.

**Comment by i-make-robots** [comment]
If you are not a native speaker, kick it sideways to someone who is. Focus on making your native tongue translate well. :)

**Comment by i-make-robots** [comment]
Should we name the language files based on https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes ?

**Reply by PPAC37** [reply]
> Should we name the language files based on https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes ?

Maybe and also try to match https://docs.oracle.com/javase/7/docs/api/java/util/ResourceBundle.html#getBundle(java.lang.String,%20java.util.Locale,%20java.lang.ClassLoader,%20java.util.ResourceBundle.Control)

`<a catalogue name>_<language_2_char_code_iso_639-1>[_<contry_2_char_code_iso_3166>[_<variant>]].xml`

so something like  : 

Makelangelo_en.xml
Makelangelo_fr.xml
Makelangelo_fr_FR.xml
Makelangelo_fr_CA.xml
...

**Comment by i-make-robots** [comment]
https://crowdin.com/ ?

**Reply by PPAC37** [reply]
is it long terme free ?

**Reply by PPAC37** [reply]
lets try :) https://crwd.in/marginallyclevermakelangelo-so

**Comment by i-make-robots** [comment]
Translator could be completely replaced with ResourceBundle and then all the work to date would be redone with all the associated headaches and bugs.  

If it is not completely replaced then we could have two competing similar systems...and that's messy.

So for those reasons I'm not a fan of ResourceBundle.  Long term maybe someone wants to do all that work.  I wish that person good luck!

**Comment by i-make-robots** [comment]
While updating documentation I noticed that `plotter settings` still shows `max speed` instead of `travel speed`.  I have changed it in english.  how do we mark the others as out of date?

**Reply by PPAC37** [reply]
> how do we mark the others as out of date?

Not easy, I have not yet found a practical technique, I had proposed to put a comment in the other files ... (like `<!-- to update / review / retranslate as value have change in english.xml -->` ... )

To be radical and simpler, if the value changes, the key should also be changed (as this key will not exist in the other translations... and therefore the translation cannot be inconsistent...)

or else need to keep  up to date a list of the values that have been changed ... but there too this does not seem to be ideal...

Without a tool to clearly identify the types of modification of english.xml (deletion / addition / renaming of keys and/or modification of value) it will remain "handmade" and difficult to manage well.

**Comment by i-make-robots** [comment]
Yeah.. the best way I could see was post it here.  All translators should
follow this thread and devs should keep tech talk to a minimum.

On Wed, Mar 2, 2022 at 12:23 PM PPAC37 ***@***.***> wrote:

> how do we mark the others as out of date?
>
> Not easy, I have not yet found a practical technique, I had proposed to
> put a comment in the other files ... (like  ... )
>
> To be radical and simpler, if the value changes, the key should also be
> changed (as this key will not exist in the other translations... and
> therefore the translation cannot be inconsistent...)
>
> or else need to keep up to date a list of the values that have been
> changed ... but there too this does not seem to be ideal...
>
> Without a tool to clearly identify the types of modification of
> english.xml (deletion / addition / renaming of keys and/or modification of
> value) it will remain "handmade" and difficult to manage well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MarginallyClever/Makelangelo-software/discussions/392#discussioncomment-2283100>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALFRBTSUDZO5KV2ON764CDU57E3BANCNFSM5KTETUMQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> <MarginallyClever/Makelangelo-software/repo-discussions/392/comments/2283100
> @github.com>
>


-- 

Dan Royer

Owner
------------------------------
1 (604) 259 9564 <16042599564> ***@***.***
www.marginallyclever.com <https://www.facebook.com/MarginallyClever/>
<https://www.instagram.com/imakerobots/>
<https://www.youtube.com/channel/UCfbRxqjuOgE2EzRKxcePArw>
<https://twitter.com/MarginallyC> <https://github.com/MarginallyClever/>

**Comment by i-make-robots** [comment]
```
	<!-- to update / review -->
	<string>
		<key>SaveGCode.splitGCodeTitle</key>
		<value>Many colors detected</value>
	</string>

	<!-- to update / review -->
	<string>
		<key>SaveGCode.splitGCodeQuestion</key>
		<value>There are %1 tool changes in this drawing.  Split into separate files?</value>
	</string>```
